File: src/day1/TwoCrystalBalls.ts
import _default from "ts-jest";

export default function two_crystal_balls(breaks: boolean[]): number {
    const jumpAmount :number = Math.floor(Math.sqrt(breaks.length));

    let i :number = jumpAmount;
    for (; i < breaks.length; i += jumpAmount) {
        if (breaks[i]) {
            break;
        }
    }

    i -= jumpAmount;

    for (let j :number = 0; j < jumpAmount && i < breaks.length; ++j, ++i) {
        if (breaks[i]) {
            return i;
        }
    }

    return -1;
}

File: go/algorithms/search/twocrystalballs.go
package search

import (
	"math"
)

func TwoCrystalBalls(breaks []bool) int {
	jumpAmount := int(math.Sqrt(float64(len(breaks))))

	i := jumpAmount
	for ; i < len(breaks); i += jumpAmount {
		if breaks[i] {
			break
		}
	}

	j := 0
	for ; j < jumpAmount && i < len(breaks); j, i = j+1, i+1 {
		if breaks[i] {
			return i
		}
	}
	return -1
}
